File: examples/factors_algeb_2.py
#
# jython examples for jas.
# $Id$
#

import sys

from java.lang import System

from jas import QQ, AN, Ring, PolyRing
from jas import terminate, startLog

# polynomial examples: factorization over Q(i)(sqrt2)

Q = PolyRing(QQ(),"i",PolyRing.lex);
print "Q     = " + str(Q);
#is automatic: Q.inject_variables();
print "i     = " + str(i);

imag = i**2 + 1;
print "imag  = " + str(imag);
Qi = AN(imag,field=True);
print "Qi    = " + str(Qi.factory());
#is automatic: Qi.inject_variables();
print "i     = " + str(i);
print;

Wr = PolyRing(Qi,"w2",PolyRing.lex)
print "Wr    = " + str(Wr);
#is automatic: Wr.inject_variables();
print "i     = " + str(i);
print "w2    = " + str(w2);

w2p = w2**2 - 2;
print "w2p   = " + str(w2p);
Qw2 = AN(w2p,field=True);
print "Qw2   = " + str(Qw2.factory());
#is automatic: Qw2.inject_variables();
print "i     = " + str(i);
print "w2    = " + str(w2);
print;

Qiw2 = PolyRing(Qw2,"x",PolyRing.lex)
print "Qiw2  = " + str(Qiw2);
#is automatic: Qiw2.inject_variables();
print "i     = " + str(i);
print "w2    = " + str(w2);
print "x     = " + str(x);
print;

#sys.exit();

f = ( x**2 + 1 ) * ( x**2 - 2 );

print "f = ", f;
print;

startLog();

t = System.currentTimeMillis();
G = Qiw2.factors(f);
t = System.currentTimeMillis() - t;
#print "G = ", G;
#print "factor time =", t, "milliseconds";

g = one;
for h, i in G.iteritems():
    print "h**i = (", h, ")**" + str(i);
    h = h**i;
    g = g*h;
#print "g = ", g;
print;

if cmp(f,g) == 0:
    print "factor time =", t, "milliseconds,", "isFactors(f,g): true" ;
else:
    print "factor time =", t, "milliseconds,", "isFactors(f,g): ",  cmp(f,g);
print;

#startLog();
terminate();

File: examples/factors_algeb_2.rb
#
# jruby examples for jas.
# $Id$
#

require "examples/jas"

# polynomial examples: factorization over Q(i)(sqrt2)

Q = PolyRing.new(QQ(),"i",PolyRing.lex);
puts "Q     = " + str(Q);
e,a = Q.gens();
puts "e     = " + str(e);
puts "a     = " + str(a);
imag = a**2 + 1;
puts "imag  = " + str(imag);
Qi = AN(imag,true);
puts "Qi    = " + str(Qi.factory());
one,i = Qi.gens();
puts "one   = " + str(one);
puts "i     = " + str(i);
puts;

Wr = PolyRing.new(Qi,"w2",PolyRing.lex)
puts "Wr    = " + str(Wr);
e,a,b = Wr.gens();
puts "e     = " + str(e);
puts "a     = " + str(a);
puts "b     = " + str(b);
w2 = b**2 - 2;
puts "w2    = " + str(w2);
Qw2 = AN(w2,true);
puts "Qw2   = " + str(Qw2.factory());
one,i,w2 = Qw2.gens();
puts "one   = " + str(one);
puts "i     = " + str(i);
puts "w2    = " + str(w2);
puts;

Qiw2 = PolyRing.new(Qw2,"x",PolyRing.lex)
puts "Qiw2  = " + str(Qiw2);

one,i,w2,x  = Qiw2.gens();
puts "one   = " + str(one);
puts "i     = " + str(i);
puts "w2    = " + str(w2);
puts "x     = " + str(x);
puts;

#sys.exit();

f = ( x**2 + 1 ) * ( x**2 - 2 );

puts "f = " + str(f);
puts;

startLog();

t = System.currentTimeMillis();
G = Qiw2.factors(f);
t = System.currentTimeMillis() - t;
#puts "G = ", G;
#puts "factor time =", t, "milliseconds";
puts;

g = one;
for h, i in G
    puts "h**i = (" + str(h) + ")**" + str(i);
    h = h**i;
    g = g*h;
end
#puts "g = ", g;
puts;

puts "factor time = " + str(t) + " milliseconds," + "isFactors(f,g): " + str(f==g);
puts;

#startLog();
terminate();
